File: 2D-Arrays/walls-and-gates.js
/** 
You are given a grid of size m x n with each cell representing one of three possible values:

-1 representing a wall.
0 representing a gate.
INF (2147483647) representing an empty room.
Fill each empty room with the distance to the nearest gate. If it is impossible to reach a gate, leave the room with an infinite distance.
*/
const EMPTY = 2147483647;
const WALL = -1;
const GATE = 0;

const directions = [
	// Up
	[-1,0],
	// Right
	[0,1],
	// Down
	[1, 0],
	// Left
	[0,-1]
];

const wallsAndGates = function(matrix) {
	for (let row = 0; row < matrix.length; row++) {
		for (let col = 0; col < matrix[0].length; col++) {
			// Find the gate and start DFS
			if (matrix[row][col] === GATE) {
				DFS(matrix, row, col, 0);
			}
		}
	}
	
	// Modify the matrix don't return else return.
	//return matrix;
}

const DFS = function(matrix, row, col, currentStep) {
	// Check the boundaries and avoid gates and currentStep should not be greater than the matrix value
	if (row < 0 ||  row >= matrix.length || col < 0 || col >= matrix[0].length || currentStep > matrix[row][col]) {
		return;
	}
	
	matrix[row][col] = currentStep;
	for (let i=0; i < directions.length; i++) {
		const currentDir = directions[i];
		DFS(matrix, row + currentDir[0], col + currentDir[1], currentStep + 1);
	}
}

File: 2D-Arrays/walls-and-gates.py
Class Solution:
    DIRECTIONS = [
        # Up
        [-1, 0],
        # Right
        [0, 1],
        # Down
        [1, 0],
        # Left
        [0, -1]
    ];
    WALL = -1
    GATE = 0
    INF = 2**31
    
    def wallsAndGates(self, grid):
        # Sequential Order to find the Gates
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] = self.GATE:
                    self.DFS(grid, i, j, 0)
        
        return grid
                    
        
        
    def DFS(self, grid, i, j, distance):
        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]):
            return
        
        if grid[i][j] == self.WALL or grid[i][j] == self.GATE:
            return
        
        if distance < grid[i][j]:
            grid[i][j] = distance
            for i in range(len(self.Directions)):
                nextRow = self.Directions[i][0] + i
                nextCol = self.Directions[i][1] + j
                self.DFS(grid,nextRow, nextCol, distance + 1)
        
        return
        
# Time Complexity using DFS (Avg) - O(mn)
# Space Complexity  O(mn)
